File: search/search_algorithms.go
package search

import "fmt"

type Node struct {
    value int
    left *Node
    right *Node
}

func NewNode (value int) *Node {
    return &Node{value, nil, nil}
}

func (node *Node) AddChildren (leftNode, rightNode *Node) {
    node.left = leftNode
    node.right = rightNode
}

func DepthFirstSearch(rootNode *Node, value int) *Node {
    fmt.Println("Visiting node ", rootNode.value)
    if rootNode.value == value {
        fmt.Println("Found value ", rootNode.value)
        return rootNode
    }

    if (rootNode.left == nil && rootNode.right == nil){
        return nil
    }

    leftBranch := DepthFirstSearch(rootNode.left, value)
    if leftBranch != nil {
        return leftBranch
    }
    return DepthFirstSearch(rootNode.right, value)
}

func main() {
    node1 := NewNode(1)
    node2 := NewNode(2)
    node3 := NewNode(3)
    node4 := NewNode(4)
    node5 := NewNode(5)
    node6 := NewNode(6)
    node7 := NewNode(7)
    node8 := NewNode(8)

    // use AddChildren

    node1.AddChildren(node2, node7)
    node2.AddChildren(node3, node4)
    node4.AddChildren(node5, node6)
    node7.AddChildren(node8, nil)

    DepthFirstSearch(node1, 4)
}

File: search/search_algorithms.py
class Node(object):
     def __init__(self, value, left=None, right=None):
         self.value = value
         self.left = left
         self.right = right
     def __repr__(self):
         return u'Node(value={})'.format(self.value)


def dfs(node, value):
     if not node:
         return None
     print 'Visiting node %d...' % node.value
     if node.value == value:
         print 'Found value %d!' % value
         return node
     if not (node.left or node.right):
         return None
     left_branch = dfs(node.left, value)
     if left_branch:
         return left_branch
     return  dfs(node.right, value)

def dfs_iterative(node, value):
     stack = [node]
     while True:
         if not stack:
             return None
         current_node = stack.pop()
         print 'Visiting node %d...' % current_node.value
         if current_node.value == value:
             print 'Found value %d!' % value
             return current_node
         if not(current_node.left or current_node.right):
             continue
         if current_node.right:
             stack.append(current_node.right)
         if current_node.left:
             stack.append(current_node.left)

def bfs_iterative(node, value):
     queue = [node]
     while queue:
         current_node = queue[0]
         print 'Visiting node %d...' % current_node.value
         if current_node.value == value:
             print 'Found value %d!' % value
             return current_node
         if not(current_node.left or current_node.right):
             queue = queue[1:]
             continue
         if current_node.left:
             queue.append(current_node.left)
         if current_node.right:
             queue.append(current_node.right)
         queue = queue[1:]
